tiledb/cloud/decoders.py

```diff
@@ -0,0 +1,60 @@
+"""Classes that know how to decode UDF results."""
+
+import abc
+import dataclasses
+import json
+from typing import Any, Generic, TypeVar
+
+import cloudpickle
+import pandas
+import pyarrow
+
+from tiledb.cloud import tiledb_cloud_error as tce
+from tiledb.cloud.rest_api import models
+
+_T = TypeVar("_T")
```

Comment:
These `TypeVars` are effectively parametrized over the three result types, right? Not sure if it's necessary, but wonder if bounding the typevar would be useful from a documentation standpoint.

Reply:
The parameter is actually intended to be the type from the user’s function. For instance, if you do

```python
def my_func(*args, **kwargs) -> MyType: ...

n = dag.submit_udf(my_func, ...)
# n is a Node[MyType]
# n._real_result should be a Result[MyType].
# n._real_result.decoder will be an AbstractDecoder[MyType].
```

Because of the way `result_format` works, this is not perfect. If `n` is serialized as a Pickle, everything works as expected. However, if `n` is serialized as Arrow, it should be a `pyarrow.Table` and if it’s JSON, it is pretty much an `Any`.

It’s…not impossible, but extremely verbose to try and get this to 100% correctness based upon all the layers of function calls with `*args` and `**kwargs` that we have going on. This 80–90% solution was at least good enough to help me out as I was implementing the internals, so I think it should provide at least *some* value to users.

Reply:
Ok... TBH this feels a little odd for python, but if the type-checker handles this then I'll trust you that it's useful 😄